common/src/stack/switch/pylib/switch/m7800.py

```diff
@@ -316,72 +318,256 @@ def del_partition_member(self, partition, guid):
+		num_steps = int(re.match(steps_regex, num_steps, flags = re.IGNORECASE).group('max_steps'))
+		completed_steps = len([result for result in results if '100.0%' in result])
+		if completed_steps != num_steps:
+			errors = self._get_expected_errors(command_response = results)
+			raise SwitchException(
+				f'Only {completed_steps} of {num_steps} firmware install steps appear to have completed successfully: {errors}'
+			)
 
 	def image_delete(self, image):
-		self.proc.say(f'image delete {image}')
+		"""Commands the switch to delete the firmware image with the provided name.
+
+		The image must have been previously loaded onto the switch with image_fetch().
+		If the deletion appears to fail, a SwitchException is raised.
+		"""
+		results = self.proc.ask(cmd = f'image delete {image}')
+		errors = self._get_errors(command_response = results)
+		if any(errors):
+			raise SwitchException(f'Image delete failed with error {errors}')
 
 
 	def image_fetch(self, url):
-		self.proc.ask(f'image fetch {url}', timeout=900)
+		"""Commands the switch to fetch a firmware image from the provided URL.
 
+		The URL must begin with one of the supported protocols or a SwitchException is raised.
+		If the transfer appears to fail, a SwitchException is raised.
+		"""
+		# validate the fetch url protocol is one we support
+		if not any((url.startswith(protocol) for protocol in self.SUPPORTED_IMAGE_FETCH_PROTOCOLS)):
+			raise SwitchException(f'Image fetch URL must be one of the following supported protocols {self.SUPPORTED_IMAGE_FETCH_PROTOCOLS}')
+
+		results = self.proc.ask(cmd = f'image fetch {url}', timeout = 900)
+		# check for success indicators and raise an error if not found.
+		if not any(('100.0%' in result for result in results)):
+			errors = self._get_expected_errors(command_response = results)
+			raise SwitchException(f'Image fetch failed with error {errors}')
+
+	def _get_installed_images(self, command_response):
+		"""Extract the currently installed images from the `show images` command response.
+
+		This attempts to identify the block of text where the installed images are listed.
+		If that block cannot be reliably found, a SwitchException is raised.
+		If no installed images are found within the installed images block, a SwitchException is raised.
+		"""
+		# strip out blank lines
+		command_response = [response for response in command_response if re.sub(r'\s+', '', response)]
+		# get the start and end indices for where we are going to look for installed images
+		start_marker = 'Installed images'
+		end_marker = 'Last boot partition'
+		block_indices_regex = fr'{start_marker}|{end_marker}'
+		block_indices = [
+			index for index, message in enumerate(command_response)
+			if re.search(block_indices_regex, message, re.IGNORECASE) is not None
+		]
+		if len(block_indices) != 2:
+			raise SwitchException(f'Ambiguous installed images block. Expected one start and one end marker: {block_indices}')
+		start_index, end_index = block_indices
+		if (
```

Comment:
> Confused about this check -- (without having run the code) shouldn't the previous regex + `if` have caught this?

This is catching if, for some reason, the `Installed images` and `Last boot partition` are in the wrong order, as `Installed images` is expected to be found before `Last boot partition` in the output. Maybe it would be clearer to re-constitute the output lines back into one mega string and use a multi-line regex to enforce this?

Reply:
I vote for it being cleaner to join back together all the output, and enforce/parse the output as one chunk. As long as the output isn't megabytes, it should be worth the extra memory for the simplification of the code.